### PR TITLE
[frontend/backend]  Marking definitions for uploaded files (#5823)

### LINKED ISSUE #5823: Marking definitions for uploaded files
## Use case

Ability to associate marking definitions to a file when uploading it from Data tab or Data/import.

Related to https://github.com/OpenCTI-Platform/opencti/issues/5797

### PR DESCRIPTION
**Pull request by ValentinBouzinFiligran**
### Proposed changes

* marking step added before file upload 
* marking form added on free text upload form
* fileMarkings key passed on global upload query
* fileMarkings key passed on entity upload query

### Related issues

* https://github.com/OpenCTI-Platform/opencti/issues/5823
*

### Checklist

- [x] I consider the submitted work as finished
- [x] I tested the code for its functionality
- [ ] I wrote test cases for the relevant uses case (coverage and e2e)
- [ ] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

* https://github.com/OpenCTI-Platform/opencti/pull/6030
* https://github.com/OpenCTI-Platform/client-python/pull/613
* https://github.com/OpenCTI-Platform/connectors/pull/2011

### CODE REVIEW

**Comment by Kedae** [comment]
File marking resolution in Content tab of an entity with file creation doesn't work

**Comment by Kedae** [comment]
Additionnaly : we are missing default file marking of entity when missing + we need to check the importFiles

File: opencti-platform/opencti-front/src/private/components/common/files/FileImportMarkingSelectionPopup.tsx [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  const { t_i18n } = useFormatter();
+
+  const handleSubmit = (values: SubmittedMarkingsType) => {
+    const fileMarkings = values.fileMarkings.map(({ value }) => value);
+    closePopup();
+    handleUpload(fileMarkings);
+  };
+
+  return (
+    <div>
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/common/files/FileImportMarkingSelectionPopup.tsx]: I think you can remove the div, there is only one child.

File: opencti-platform/opencti-front/src/private/components/common/files/FileUploader.tsx [hunk 2]
```diff
@@ -9,35 +9,36 @@ import { commitMutation, MESSAGING$ } from '../../../../relay/environment';
...
 const fileUploaderGlobalMutation = graphql`
-  mutation FileUploaderGlobalMutation($file: Upload!) {
-    uploadImport(file: $file) {
-      id
-      ...FileLine_file
+    mutation FileUploaderGlobalMutation($file: Upload!, $fileMarkings: [String]) {
+        uploadImport(file: $file, fileMarkings: $fileMarkings) {
+            id
+            ...FileLine_file
+        }
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/common/files/FileUploader.tsx]: why the indentation change?

File: opencti-platform/opencti-front/src/private/components/common/files/FileUploader.tsx [hunk 3]
```diff
@@ -9,35 +9,36 @@ import { commitMutation, MESSAGING$ } from '../../../../relay/environment';
...
-        id
-        ...FileLine_file
-        metaData {
-          entity {
-            ... on StixObject {
-              id
-            }
-            ... on StixDomainObject {
-              ...PictureManagementViewer_entity
+    mutation FileUploaderEntityMutation($id: ID!, $file: Upload!, $fileMarkings: [String]) {
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/common/files/FileUploader.tsx]: same. 
Seems like your code formatter config uses 4 spaces for gql

**Comment by labo-flg** [comment]
I've done some tests with a Malware and the Stix json export.
I managed to get one export with content + file markings to TLP;GREEN but when I try with TLP;RED I get

`line 1980, in prepare_export\n    if stix_core_relationship[\"to\"][\"id\"] != entity[\"x_opencti_id\"]\nTypeError: 'NoneType' object is not subscriptable`

File: opencti-platform/opencti-front/lang/front/de.json [hunk 4]
```diff
@@ -2102,6 +2102,7 @@
   "Select all nodes": "Alle Knoten auswählen",
   "Select by entity type": "Auswahl nach Entitätstyp",
   "Select destination": "Ziel auswählen",
+  "Select file marking définitions": "Definitionen der Dateikennzeichnung auswählen",
```
Comment (aHenryJard) [inline opencti-platform/opencti-front/lang/front/de.json]:
```suggestion
  "Select file marking definitions": "Definitionen der Dateikennzeichnung auswählen",
```
No "é" in English

File: opencti-platform/opencti-front/src/private/components/common/files/FileImportMarkingSelectionPopup.tsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        }}
+        onSubmit={handleSubmit}
+      >
+        {({ resetForm, submitForm }) => (
+          <Dialog open={isOpen} fullWidth={true} PaperProps={{ elevation: 1 }} onClose={() => {
+            resetForm();
+            closePopup();
+          }}
+          >
+            <DialogTitle>{t_i18n('Select file marking définitions')}</DialogTitle>
```
Comment (aHenryJard) [inline opencti-platform/opencti-front/src/private/components/common/files/FileImportMarkingSelectionPopup.tsx]:
```suggestion
            <DialogTitle>{t_i18n('Select file marking definitions')}</DialogTitle>
```

File: opencti-platform/opencti-front/lang/front/zh.json [hunk 6]
```diff
@@ -2102,6 +2102,7 @@
   "Select all nodes": "选择所有节点",
   "Select by entity type": "按实体类型选择",
   "Select destination": "选择目的地",
+  "Select file marking définitions": "选择文件标识定义",
```
Comment (aHenryJard) [inline opencti-platform/opencti-front/lang/front/zh.json]:
```suggestion
  "Select file marking definitions": "选择文件标识定义",
```

File: opencti-platform/opencti-front/lang/front/ja.json [hunk 7]
```diff
@@ -2102,6 +2102,7 @@
   "Select all nodes": "すべてのノードを選択",
   "Select by entity type": "エンティティ種別を指定して選択",
   "Select destination": "送信先の選択",
+  "Select file marking définitions": "ファイルマーキングの定義を選択",
```
Comment (aHenryJard) [inline opencti-platform/opencti-front/lang/front/ja.json]:
```suggestion
  "Select file marking definitions": "ファイルマーキングの定義を選択",
```

File: opencti-platform/opencti-front/lang/front/fr.json [hunk 8]
```diff
@@ -2102,6 +2102,7 @@
   "Select all nodes": "Sélectionner tous les noeuds",
   "Select by entity type": "Sélectionner par type",
   "Select destination": "Sélectionner la destination",
+  "Select file marking définitions": "Sélectionner les définitions de marquage de fichiers",
```
Comment (aHenryJard) [inline opencti-platform/opencti-front/lang/front/fr.json]:
```suggestion
  "Select file marking definitions": "Sélectionner les définitions de marquage de fichiers",
```

File: opencti-platform/opencti-front/lang/front/es.json [hunk 9]
```diff
@@ -2102,6 +2102,7 @@
   "Select all nodes": "Séleccionar todos los nodos",
   "Select by entity type": "Séleccionar por tipo",
   "Select destination": "Seleccionar destino",
+  "Select file marking définitions": "Seleccionar definiciones de marcado de archivos",
```
Comment (aHenryJard) [inline opencti-platform/opencti-front/lang/front/es.json]:
```suggestion
  "Select file marking definitions": "Seleccionar definiciones de marcado de archivos",
```

File: opencti-platform/opencti-front/lang/front/en.json [hunk 10]
```diff
@@ -2102,6 +2102,7 @@
   "Select all nodes": "Select all nodes",
   "Select by entity type": "Select by entity type",
   "Select destination": "Select destination",
+  "Select file marking définitions": "Select file marking définitions",
```
Comment (aHenryJard) [inline opencti-platform/opencti-front/lang/front/en.json]:
```suggestion
  "Select file marking definitions": "Select file marking définitions",
```

File: opencti-platform/opencti-graphql/src/resolvers/file.js [hunk 11]
```diff
@@ -28,7 +28,7 @@ const fileResolvers = {
     creator: (metadata, _, context) => creatorLoader.load(metadata.creator_id, context, context.user),
   },
   Mutation: {
-    uploadImport: (_, { file }, context) => uploadImport(context, context.user, file),
+    uploadImport: (_, args, context) => uploadImport(context, context.user, args),
```
Comment (aHenryJard) [inline opencti-platform/opencti-graphql/src/resolvers/file.js:31]: Have you checked that this code still works fine ?
https://github.com/OpenCTI-Platform/client-python/blob/65794bb194e9e87411d735f630308cdbeff4332b/pycti/api/opencti_api_client.py#L626-L634

File: opencti-platform/opencti-graphql/src/domain/stixCoreObject.js [hunk 12]
```diff
@@ -406,7 +406,7 @@ export const stixCoreObjectImportPush = async (context, user, id, file, args = {
       const key = `${filePath}/${filename}`;
       meta.external_reference_id = generateStandardId(ENTITY_TYPE_EXTERNAL_REFERENCE, { url: `/storage/get/${key}` });
     }
-    const { upload: up, untouched } = await upload(context, user, filePath, file, { meta, noTriggerImport, entity: previous });
+    const { upload: up, untouched } = await upload(context, user, filePath, file, { meta, noTriggerImport, entity: previous, file_markings });
```
Comment (aHenryJard) [inline opencti-platform/opencti-graphql/src/domain/stixCoreObject.js:409]: I don't see any change on '../database/file-storage' shouldn't marking be stored in elastic with indexFileToDocument call there ?

**Review by aHenryJard — COMMENTED** [review]
Since you are changing uploadImport mutation, it could be great to add at least one resolver test that covers it.

File: opencti-platform/opencti-graphql/src/resolvers/file.js (hunk 11 above)
Comment (labo-flg) [reply opencti-platform/opencti-graphql/src/resolvers/file.js:31]: Can't hurt to check, but API is unchanged (we still expect an object args).

**Comment by labo-flg** [comment]
I've tested the upload, access check is OK.

File: opencti-platform/opencti-graphql/src/resolvers/file.js (hunk 11 above)
Comment (labo-flg) [reply opencti-platform/opencti-graphql/src/resolvers/file.js:31]: However, we might want to update the python client to provide the new parameter `file_markings` if we want to align with the gql API.

File: opencti-platform/opencti-graphql/src/domain/stixCoreObject.js (hunk 12 above)
Comment (labo-flg) [reply opencti-platform/opencti-graphql/src/domain/stixCoreObject.js:409]: @aHenryJard it's already taken care of in a previous PR ; it's stored in elastic properly

File: opencti-platform/opencti-graphql/src/resolvers/file.js (hunk 11 above)
Comment (labo-flg) [reply opencti-platform/opencti-graphql/src/resolvers/file.js:31]: I'll do it in another PR, let's move on with this one

**Review by labo-flg — APPROVED** [review]

**Comment by labo-flg** [comment]
> Since you are changing uploadImport mutation, it could be great to add at least one resolver test that covers it.


There is no resolver tests for file domain yet.

This would require significant work, let's split this in another PR.

**Pull request merged**